man/putrec.3
.TH PUTREC 3 "12 September 1995"
.FB
.SH NAME
putrec \- write current record to database
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_putrec(rec, db)
.br
long rec;
.br
fb_database *db;
.SH DESCRIPTION
As explained in the \fIdatabase\fP(4)
manual page, a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
\fIGetrec\fP(3) or \fIgetxrec\fP(3)
are used to retrieve an existing database
record from the database into the array of field descriptors.
Then, these field structures may be updated by using
the \fIstore\fP(3) routine. Once a record has been modified,
\fIfb_putrec\fP is called to physically
write the fields of data as a single record to the database.
.PP
\fIFb_putrec\fP
stores logical record \fIrec\fP into the database pointed to by \fIdb\fP.
The map file descriptor stored in \fIdb->mfd\fP is used to determine the old
location and length of the record within the database file
described by the file descriptor stored in \fIdb->fd\fP.
.PP
\fIGetrec\fP(3)
will place the current record number in a field inside
of the database structure called \fIdb->rec\fP when a record is read.
Most times this will be the
same number used as a parameter to \fIfb_putrec\fP.
.PP
The records in any \fIFirstBase\fP database are indexed starting at
physical record
number one (1). When using \fIfb_putrec\fP,
it is important that the record number \fIrec\fP
represents a record already in the database, or one more than the database
holds. I.E., the range that \fIrec\fP can take on is between 1 and
the database record count + 1. (1 <= \fIrec\fP <= \fIdb->reccnt\fP + 1).
In other words, new records must be assigned the next sequential
record number in the database.
.PP 
If the record did not exist before, or if the record will not fit in the
place it did exist, a new storage area is located and the old area
is linked into a free storage list.
New storage areas come from the free list or are allocated at
the end of the database.
.PP
\fIFb_putrec\fP
uses the database field descriptors to gather and write the fields
to the database. These field descriptors will point into one of the
buffers in the database structure, \fIdb->orec\fP
or \fIdb->arec\fP depending on whether the field has been altered or not.
.PP
Normally, \fIfb_putrec\fP(3) will store records in a variable length format.
However, some applications that modify the same records over and over
might get better performance with fixed width records. By setting the
external short integer variable cdb_fixedwidth to one (1) before calling
\fIfb_opendb\fP, or by setting the short integer \fIfixedwidth\fP in the
database structure after the call to \fIfb_opendb\fP,
\fIfb_putrec\fP will pad the end of each record to create fixed width
entries. Alternately, system wide fixed width \fIFirstBase\fP
record storage can be defined via \fIsetup\fP(5).
.SH CLIENT/SERVER
The \fIfb_putrec\fP routine will transparently
call the \fIfb_putrec_clnt\fP mechanism
when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP is set to one.
See \fIfb_clnt_create\fP(3) for more details.
.SH SEE ALSO
addrec(3), getrec(3), getxrec(3), opendb(3), store(3), database(4), setup(5).
.SH DIAGNOSTICS
\fIFb_putrec\fP will return the error code FB_ERROR if any of its reads,
writes, or seeks fail, or if \fIrec\fP <= 0 or rec > db->reccnt + 1.
An FB_AOK signal is returned on correct completion.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/putrec.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: putrec.c,v 9.0 2001/01/09 02:57:01 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Putrec_sid[] = "@(#) $Id: putrec.c,v 9.0 2001/01/09 02:57:01 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

#if FB_PROTOTYPES
static long countrec(fb_database *hp);
static long p_putrec(fb_database *hp, long size, long rlen);
#else /* FB_PROTOTYPES */
static long countrec();
static long p_putrec();
#endif /* FB_PROTOTYPES */

extern short int cdb_use_rpc;
extern short int cdb_returnerror;
extern short int cdb_error;
extern char cdb_EOREC;
extern char cdb_FILLE;
extern char cdb_FILLC;

extern long fb_getfree(int fd, int mfd, long size, long *rlen);
static long countrec(fb_database *hp);

/* 
 * putrec - put hp->orec/hp->arec to logical record n of hp->fd.
 *
 */

   fb_putrec(n, hp)
      long n;
      fb_database *hp;
   
      {
         long avail, freep, rpos, rlen, npos, frlen;
         long wlen, plen, size;

#if RPC
         if (cdb_use_rpc)
            return(fb_putrec_clnt(n, hp));
#endif /* RPC */
         if (n <= 0 || n > hp->reccnt + 1)
            return(FB_ERROR);
         hp->rec = n;
	 size = countrec(hp);
         /* set projected length */
         if (!hp->fixedwidth)
	    plen = size;
         else
            plen = hp->recsiz + 1;
	 rlen = 0L;
	 if (fb_getmap(hp->mfd, n, &avail, &freep, &rpos, &rlen) == FB_ERROR){
	    fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_MAP], hp->dmap);
	    return(FB_ERROR);
	    }
	 if (size <= rlen && plen <= rlen){ /* replace record - it fits here */
	    if (lseek(hp->fd, rpos, 0) < 0L){
	       fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_DATA], hp->dbase);
	       return(FB_ERROR);
	       }

	    /* no reclaiming here -- once rlen, always rlen, or new! */

	    if (p_putrec(hp, size, rlen) == FB_ERROR)
               return(FB_ERROR);
	    }
	 else {		/* must find new place for record */
	    /* 
	     *  the idea here is to find a free space, or use avail.
	     *  if a free space exists, do not changes its rlen.
	     */
	    
	    frlen = rlen;
            npos = fb_getfree(hp->fd, hp->mfd, size, &rlen);
	    if (lseek(hp->fd, npos, 0) < 0L){	/* put the record there! */
	       fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_DATA], NIL);
	       return(FB_ERROR);
	       }
	    wlen = p_putrec(hp, size, rlen);	/* rlen may change */
            if (wlen == FB_ERROR)
               return(FB_ERROR);
	    if (rlen == 0L)
	       avail += wlen;			/* bump avail to past cur. */
            if (rpos > 0)			/* free up old space */
	       if (fb_putfree(hp->fd, hp->mfd, rpos, frlen, hp) == FB_ERROR){
                  cdb_error = FB_WRITE_ERROR;
	          fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_DATA],
                     SYSMSG[S_BAD_MAP]);
		  return(FB_ERROR);
		  }
	    if (fb_putmap(hp->mfd, n, avail, npos, wlen)==FB_ERROR){
               cdb_error = FB_WRITE_ERROR;
	       fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_DATA],SYSMSG[S_BAD_MAP]);
	       return(FB_ERROR);
	       }
	    }
	 return(FB_AOK);
      }

/*
 * countrec - count the characters in the record by looking at all
 *	fields. include the NULL markers. 
 *	assume record length can fit in sizeof(long).
 */
 
   static long countrec(hp)
      fb_database *hp;
      
      {
	 register int i;
	 fb_field **kf;
	 long size;
	 
	 kf = hp->kp;
	 for (size = 0L, i = 0; i <= hp->nfields; i++, kf++){
            if ((*kf)->type != FB_BINARY){
               if ((*kf)->fld == NULL)
                  (*kf)->fld = NIL;			/* just for safety */
               if ((*kf)->type != FB_FORMULA && (*kf)->dflink == NULL &&
                     (*kf)->type != FB_LINK)
                  size += strlen((*kf)->fld) + 1L;	/* one for NULL */
               }
            else
	       size += (*kf)->size;			/* FB_BINARY only */
	    }
	 return(size+1L);				/* one for cdb_EOREC */
      }

/*
 * p_putrec - put the record out to hp->fd. hp->fd is already set.
 *           follow the fb_field pointers, add FILLCs and an cdb_EOREC.
 */
 
   static long p_putrec(hp, size, rlen)
      fb_database *hp;
      long rlen, size;
      
      {
	 register char *bufp, *fldp;
	 fb_field **kf;
         register int i;
	 char *startp;
	 
	 kf = hp->kp;
	 bufp = hp->arec;
	 /* copy each fb_field and its NULL pointer to arec */
	 for (i = 0; i <= hp->nfields; i++, kf++)
	    if ((*kf)->type != FB_FORMULA && (*kf)->dflink == NULL &&
                   (*kf)->type != FB_LINK){
	       /* save beginning of fb_field position */
	       startp = bufp;
               if ((*kf)->type != FB_BINARY){	/* NULL terminated copy */
                  for (fldp = (*kf)->fld; ; fldp++){
                     *bufp++ = *fldp;
                     if (!*fldp)
                        break;
                     }
                  }
               else{				/* FB_BINARY byte copy */
                  memcpy(bufp, (*kf)->fld, (unsigned) (*kf)->size);
                  bufp += (*kf)->size;
                  }
	       (*kf)->fld = startp;
	       }
	 *bufp++ = cdb_EOREC;		 	/* mark record end */
	 rlen = MAX( size, MAX(rlen, (long) (FB_MAPREC_SIZE + 2L)));
         if (hp->fixedwidth)
            rlen = MAX(rlen, hp->recsiz + 1);
	 if (size < rlen){			/* fill hole if needed */
	    while (++size < rlen)
	       *bufp++ = cdb_FILLC;
	    *bufp++ = cdb_FILLE;		/* rec[rlen] = fille */
	    }
	 if (write(hp->fd, hp->arec, (unsigned) rlen) != rlen){
            if (cdb_returnerror){
               cdb_error = FB_WRITE_ERROR;
	       fb_lerror(FB_FATAL_PUTREC,
                  SYSMSG[S_BAD_DATA],(char *) &(hp->rec));
               return(FB_ERROR);
               }
	    fb_xerror(FB_FATAL_PUTREC, SYSMSG[S_BAD_DATA],(char *) &(hp->rec));
            }
	 /* swap arec and orec -- so multiple fb_putrec() calls will work */
	 bufp = hp->arec;
	 hp->arec = hp->orec;
	 hp->orec = bufp;
	 return(rlen);
      }
